src/ucp/rma/amo_sw.c

```diff
@@ -194,39 +194,182 @@ DEFINE_AMO_SW_OP(64)
+
+    ucp_amo_sw_loopback_send_reply(worker, flush_ctx);
+    ucp_request_put(req);
+}
+
+static ucs_status_t
+ucp_amo_sw_loopback_post(ucp_worker_h worker, ucp_atomic_req_hdr_t *amo_req_hdr)
+{
+    ucp_atomic_flush_ctx_t *flush_ctx;
+    ucp_mem_map_params_t params;
+    void *rkey_buffer;
+    size_t rkey_buffer_size;
+    void *request;
+    ucs_status_t status, status_unmap;
+
+    /* TODO: optimization: can use `am_flags & UCT_CB_PARAM_FLAG_DESC` to avoid
+     *       extra allocation and copy for some transports */
+    flush_ctx = ucs_malloc(sizeof(*flush_ctx), "loop back amo flush ctx");
+    if (flush_ctx == NULL) {
+        ucs_error("failed to allocate loopback AMO flush context");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    flush_ctx->req_hdr = *amo_req_hdr;
+    flush_ctx->value   = *(uint64_t *)(amo_req_hdr + 1);
+
+    params.field_mask  = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |
+                         UCP_MEM_MAP_PARAM_FIELD_LENGTH  |
+                         UCP_MEM_MAP_PARAM_FIELD_FLAGS;
+    params.address     = (void *)amo_req_hdr->address;
+    params.length      = amo_req_hdr->length;
+    params.flags       = 0;
+
+    /*
+     * TODO: rkey can be passed as a part of AM to avoid
+     *       mem_map/rkey_pack/rkey_pack.
+     * NOTE: if rcache is enabled then registration should be "for free" since
+     *       destination memory has been registered.
+     */
+    status = ucp_mem_map(worker->context, &params, &flush_ctx->memh);
+    if (status != UCS_OK) {
+        ucs_error("failed to map memory region %p length %zu status %s",
+                  params.address, params.length, ucs_status_string(status));
+        goto err_free_ctx;
+    }
+
+    status = ucp_rkey_pack(worker->context, flush_ctx->memh, &rkey_buffer,
+                           &rkey_buffer_size);
+    if (status != UCS_OK) {
+        ucs_error("failed to pack rkey for region %p length %zu, status %s",
+                  params.address, params.length, ucs_status_string(status));
+        goto err_unmap;
+    }
+
+    status = ucp_ep_rkey_unpack(worker->atomic_ep, rkey_buffer,
+                                &flush_ctx->rkey);
+    ucp_rkey_buffer_release(rkey_buffer);
+    if (status != UCS_OK) {
+        ucs_error("failed to unpack rkey for region %p length %zu, status %s",
+                  params.address, params.length, ucs_status_string(status));
+        goto err_unmap;
+    }
+
+    status = ucp_atomic_post(worker->atomic_ep,
+                             (ucp_atomic_post_op_t)amo_req_hdr->opcode,
+                             flush_ctx->value, amo_req_hdr->length,
+                             amo_req_hdr->address, flush_ctx->rkey);
+    if (status != UCS_OK) {
+        ucs_error("failed to post loopback AMO to ep %p, status %s",
+                  worker->atomic_ep, ucs_status_string(status));
+        goto err_unmap;
+    }
+
+    request = ucp_ep_flush_internal(worker->atomic_ep, UCT_FLUSH_FLAG_LOCAL,
```

Comment:
why need to flush? uct atomic operation has a completion callback, can use it to send the sw rma reply.

Reply:
I used ucp ep to avoid another implementation of handling lack of resources, pending, etc

Reply:
so can we use the callback on ucp atomic operation instead of issuing a flush?

Reply:
do you mean add a wrapper for 
```
    ucp_amo_init_post(req, ep, ucp_uct_op_table[opcode], op_size, remote_addr,
                      rkey, value, rkey->cache.amo_proto);

    status_p = ucp_rma_send_request_cb(req, (ucp_send_callback_t)ucs_empty_function);
```
and pass CB to `ucp_rma_send_request_cb`?

Reply:
yes